From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mediang <104988785+Mediang@users.noreply.github.com>
Date: Thu, 5 May 2022 18:24:35 +0000
Subject: [PATCH] Bromine config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 743a4091e06015b303d1eebc3b6bd393f98e4b9f..a33f2ff0745b5a62c4b2fc4140664c57841cdcf2 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1900,6 +1900,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Bromine start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getBromineConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Bromine end
+
         /**
          * Sends the component to the player
          *
